### ISSUE TITLE
[question] Hexagonal architecture and CQRS

### ISSUE THREAD

**Issue by wooyounggggg** [caffeine-library/Domain-Driven-Design#17]
## 연관 챕터
챕터 무관

## 조사 내용
CQRS를 deep dive하는 중입니다. ChatGPT에 다음과 같은 질답을 해보았습니다.

## Q. Are hexagonal architecture and cqrs necessarily coexisting?

Hexagonal architecture and CQRS (Command Query Responsibility Segregation) are not necessarily coexisting, but they can complement each other well.

Hexagonal architecture is a software design pattern that aims to create a clear separation between the business logic of an application and the infrastructure that supports it. It does this by creating a "core" domain model that is independent of the infrastructure, and by creating adapters that connect the domain model to the infrastructure.

CQRS is a pattern that separates the responsibilities of querying data (reading) and modifying data (writing) into separate components.

Hexagonal architecture can help to implement the CQRS pattern by providing a clear separation between the domain model and the infrastructure. The core domain model is responsible for the business logic, and the adapters are responsible for connecting the domain model to the infrastructure. By separating the domain model from the infrastructure, it becomes possible to optimize the read and write operations separately.

On the other hand, CQRS can help to implement the hexagonal architecture by providing a clear separation between the read and write operations. This allows you to create a clear separation between the business logic and the infrastructure, and it also makes it easier to test the business logic in isolation.

So in conclusion, while Hexagonal architecture and CQRS are not necessarily coexisting, they can be used together to create a clear separation between the business logic and the infrastructure, and between the read and write operations.

---
위와 같은 질문을 한 이유는, 사이드 프로젝트에서 Hexagonal Architecture를 적용해보면서 복잡한 조회 로직을 구현하는 것이 힘들었는데
근본적인 이유가 Command / Query model이 분리되어 있지 않았기 때문이 아니었을까란 생각이 들었기 때문입니다.

답변에도 나와있듯, 당연히 Hexagonal Architecture와 CQRS를 항상 병행할 필요는 없지만 상호 보완할 수 있는 부분임은 틀림이 없는 것 같은데요, 여러분이 경험한 프로젝트에서는 이러한 방법론들을 어떻게 적용하셔서 풀어나가고 있는지 경험도 공유해주시면 좋을 것 같습니다.

**Comment by JasonYoo1995** [comment]
- 헥사고날 아키텍처 : https://nexon-note.notion.site/3224517afeab42c0a25398886d180905

**Comment by JasonYoo1995** [comment]
Hexagonal Architecture에서는 읽기 전용 Use Case와 쓰기 전용 Use Case를 분리해서 관리할 수 있는데
이 점이 CQRS에서 읽기과 쓰기를 분리한다는 점과 일치하는 부분이 있는 것 같습니다.
